source/CollisionSet.cpp

```diff
@@ -332,23 +360,26 @@ const vector<Body *> &CollisionSet::Circle(const Point &center, double radius) c
 const vector<Body *> &CollisionSet::Ring(const Point &center, double inner, double outer) const
 {
 	// Calculate the range of (x, y) grid coordinates this ring covers.
-	int minX = static_cast<int>(center.X() - outer) >> SHIFT;
-	int minY = static_cast<int>(center.Y() - outer) >> SHIFT;
-	int maxX = static_cast<int>(center.X() + outer) >> SHIFT;
-	int maxY = static_cast<int>(center.Y() + outer) >> SHIFT;
+	const int minX = static_cast<int>(center.X() - outer) >> SHIFT;
+	const int minY = static_cast<int>(center.Y() - outer) >> SHIFT;
+	const int maxX = static_cast<int>(center.X() + outer) >> SHIFT;
+	const int maxY = static_cast<int>(center.Y() + outer) >> SHIFT;
+
+	for(auto &entry : sorted)
+	{
+		entry.isSeen = false;
+	}
```

Comment:
Doesn't this suffer from the same problem as having just a vector when you have a lot of bodies? Maybe you can combine the two approaches: Use the `seen` vector again but keep the `isShared` checks so that the entries in the vector are kept to a minimum? What do you think?

Reply:
> Doesn't this suffer from the same problem as having just a vector when you have a lot of bodies?

Is a bit different: using `seen` as a a separate vector means it needs to be loaded from memory (which could be "distant" in cache as the cache is filled with `sorted`...) while here the vector of sorted is going to be used anyway. That's the reason to store that information inside each `Entry`: even if you needs to iterate over it to reinitialize it, that memory is likely to be used again in the main loop. So the time it takes to be fetched is not wasted because just after that you read from it.

> Maybe you can combine the two approaches: Use the seen vector again but keep the isShared checks so that the entries in the vector are kept to a minimum?

That is just dropping [this commit](https://github.com/endless-sky/endless-sky/pull/7743/commits/4a66dfb6d2f1a3c774b920ef989d4cda9e61ed0b), which shown a perfomance gain for the above reason.

However I can try dropping [that commit](https://github.com/endless-sky/endless-sky/pull/7743/commits/01ab97c17d6160d8a76dd9ce07370d7639d9f100): this way there's no needs to make that loop.

NOTE: I did all the tests in battles at Mesuket, but during the Wanderer campaign. And this time waiting for a bigger set of ships, so I would trust the profiling (though it could be improved...).

Reply:
> That is just dropping [this commit](https://github.com/endless-sky/endless-sky/pull/7743/commits/4a66dfb6d2f1a3c774b920ef989d4cda9e61ed0b)

That commit does something a bit different, it still fills a vector with values the same size as the Body count. I was referring the initial commit you had, where you filled the vector with Body pointers and searched it for matches.

Reply:
I dropped the commit where the `find` was done: I agreed that it was slow than what @wjp did.

And I rebased so "style commits" go first and every commit I did show a little perfomance gain, I'm sorry I didn't mention that...

Reply:
[This was me really tired](https://github.com/endless-sky/endless-sky/pull/7743#discussion_r1039057205): please forgive me...
The commit you ask for is [this](https://github.com/endless-sky/endless-sky/pull/7743/commits/46826a5c281723582725dc71fb542fa86bab55ca), just after I made the `seen` a member

Reply:
Done, I think... :sweat_smile: